design/anonymousTokens.md

```diff
@@ -0,0 +1,96 @@
+  (e.g. `***ANONYMOUS***-[remainder of token goes here]`)
+* ATs would be created, introspected, and revoked with separate APIs from normal tokens.
+  * Token creation requires no authentication.
+* The administration UI would have AT and user lookup functionality and the ability
+  to revoke said tokens.
+* The administration UI would have the ability to set the lifetime of ATs. The default lifetime
+  could be fairly long since ATs grant few privileges.
+* All endpoints except for the AT specific endpoints would respond with an error when
+  presented with an AT.
+  
+## Interactions with KBase infrastructure
+
+* Most KBase core services (including data services) should never see an an AT.
+* Narrative:
+  * When a user without a KBase cookie in the browser loads the KBase UI, an anonymous token would
+    be generated and stored in a cookie (potentially the same cookie).
+  * Starting a narrative would assign the docker container to the username in the AT.
+    * Narratives started in this manner would always be read-only to the user and must
+      be public.
+    * The Narrative backend would not pass the AT on to any services.
+  * The UI should show that the user is in anonymous mode, their transient username, and allow
+    them to log in normally.
+* Downloads
+  * Note that DLs are significantly more work but are orthogonal to the Narrative work.
+  * JSON DLs should work as is as long as the front end does not pass the AT to DIE.
+  * Other downloads
+    * The NJS would need to understand anonymous tokens (to allow for job
+      tracking purposes) and have the ability to start jobs without a token.
+      * The NJS or catalog could have a list of apps that may be started without a
+        token - these would presumably be DL apps only.
+    * The SDK would have to support starting apps without a token.
+    * DL apps would have to support running without a token.
+      * For non-staging service apps, the Shock node created by the app would need
+        to be set public for anonymous users.
+    * The DIE should work normally for non-staging area DLs as long as the front
+      end does not pass the AT to DIE.
+    * Alternatively, non-staging area DLs could be deprecated and removed.
+      * In this case, the DIE functionality for DLing JSON could probably be
+        moved into a different service (NarrativeService?) and DIE retired.
+    * The Staging service would need to understand ATs and place folders
+      for anonymous users in a separate namespace.
+      * The Staging service should not allow anonymous uploads.
+    * Note that all of these changes could be made and deployed piecemeal.
+
+## Resourcing
+
+* Estimated effort is ~1 week for a developer familiar with the auth codebase.
+  * Requires a lot of testing to ensure anonymous tokens don't interact negatively with the
+    rest of the service.
+
+## Side Nodes
+
+* It is really weird that staging area downloads wind up as narrative cells. That seems to
```

Comment:
+1

I'd like to see a separate app mechanism (if we want to stick to "apps" here) for moving data from Workspace -> staging. It doesn't affect the user's science, it clutters the Narrative, and it slows the Narrative backend on start when it looks up state of all jobs.